contracts/JumpBall.cdc

```diff
@@ -0,0 +1,379 @@
+            }
+
+            JumpBall.games[gameID] <-! create Game(
+                id: gameID,
+                creator: self.address,
+                startTime: startTime,
+                gameDuration: gameDuration,
+                selectedStatistic: selectedStatistic,
+                creatorCap: self.collectionCap
+            )
+
+            JumpBall.addGameForUser(user: self.address, gameID: gameID)
+
+            emit GameCreated(gameID: gameID, creator: self.address, startTime: startTime)
+            return gameID
+        }
+
+        // Add an opponent to a game
+        access(all) fun addOpponent(gameID: String) {
+            let gameRef = &JumpBall.games[gameID] as &Game?
+                    ?? panic("Game does not exist.")
+
+            if !self.collectionCap.check() {
+                panic("Player does not have a valid collection capability to join a game.")
+            }
+
+            gameRef.setOpponent(opponentAddress: self.address, collectionCap: self.collectionCap)
+            JumpBall.addGameForUser(user: self.address, gameID: gameID)
+
+            emit OpponentAdded(gameID: gameID, opponent: self.address)
+        }
+    }
+
+    // Resource to manage game-specific data
+    access(all) resource Game {
+        access(all) let id: String
+        access(all) let creator: Address
+        access(all) var opponent: Address? // Opponent can be added later
+        access(all) let startTime: UFix64
+        access(all) let gameDuration: UFix64
+        access(self) var selectedStatistic: String
+        access(self) var nfts: @{UInt64: {NonFungibleToken.NFT}}
+        access(self) var ownership: {UInt64: Address}
+        access(self) var metadata: {String: AnyStruct}
+        access(self) let creatorCap: Capability<&{NonFungibleToken.Collection}>
+        access(self) var opponentCap: Capability<&{NonFungibleToken.Collection}>?
+
+        init(id: String, creator: Address, startTime: UFix64, gameDuration: UFix64, selectedStatistic: String, creatorCap: Capability<&{NonFungibleToken.Collection}>) {
+            self.id = id
+            self.creator = creator
+            self.opponent = nil
+            self.startTime = startTime
+            self.gameDuration = gameDuration
+            self.selectedStatistic = selectedStatistic
+            self.nfts <- {}
+            self.ownership = {}
+            self.metadata = {}
+            self.creatorCap = creatorCap
+            self.opponentCap = nil
+        }
+
+        // Getter for creatorCap
+        access(all) fun getCreatorCap(): Capability<&{NonFungibleToken.Collection}> {
+            return self.creatorCap
+        }
+
+        // Getter for opponentCap
+        access(all) fun getOpponentCap(): Capability<&{NonFungibleToken.Collection}>? {
+            return self.opponentCap
+        }
+
+        access(all) fun setMetadata(key: String, value: AnyStruct) {
```

Comment:
We should add an entitlement for `setMetadata` and other methods that should not be callable publicly after getting a reference to a Game resource using the `getGame` contract method